metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala

```diff
@@ -0,0 +1,167 @@
+package scala.meta.internal.metals.testProvider
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.Executors
+
+import scala.collection.concurrent.TrieMap
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext
+
+import scala.meta.internal.metals.BuildTargets
+import scala.meta.internal.metals.ClientConfiguration
+import scala.meta.internal.metals.DefinitionProvider
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.debug.BuildTargetClasses
+import scala.meta.internal.metals.debug.BuildTargetClasses.FullyQualifiedClassName
+import scala.meta.internal.metals.debug.BuildTargetClasses.Symbol
+
+import ch.epfl.scala.bsp4j.BuildTarget
+
+private final case class SymbolsPerTarget(
+    target: BuildTarget,
+    testSymbols: TrieMap[Symbol, FullyQualifiedClassName]
+)
+
+final class TestSuitesProvider(
+    buildTargets: BuildTargets,
+    buildTargetClasses: BuildTargetClasses,
+    definitionProvider: DefinitionProvider,
+    clientConfig: ClientConfiguration
+) {
+  private val executor =
+    ExecutionContext.fromExecutorService(Executors.newSingleThreadExecutor())
+
+  /**
+   * Cached, already discovered test suites.
+   * Cache is updated via refreshTestSuites method called after compilation is finished.
+   * There is no need to have synchronization since all access is available via
+   * single threaded executor.
+   *
+   * For every test suite a.MunitTestSuite BSP returns 2 symbols. However, only one of them
+   * has location in source code and it's useful.
+   * That's why FullyQualifiedClassName is used as a key.
+   * "a/MunitTestSuite., a.MunitTestSuite"
+   * "a/MunitTestSuite#, a.MunitTestSuite"
+   */
+  private val cachedTestSuites =
+    mutable.Map[FullyQualifiedClassName, TestEntry]()
+
+  /**
+   * Retrieves cached test suites.
+   */
+  def findTestSuites()
+      : CompletableFuture[java.util.List[TestSuiteDiscoveryResult]] =
+    CompletableFuture.supplyAsync(
```

Comment:
Is this needed? Can we instead use TrieMap and handle this normally?

Reply:
I used it primarily in order to avoid parallel updating of cache and it's not about synchronization, it's about duplicating work but `BatchedFunction` solves this. 
The secondary goal was to ensure that `refresh` will happen always after the end of the cache update. A situation like below could happen and test explorer won't have all of the test suites. However, I think it's not a huge issue since refresh happens very often and I already have a plan for switching approach from pull-based to push-based.
```shell
TestSuitesProvider.scala:96 "start": "start"
TestSuitesProvider.scala:56 "refresh": "refresh"
TestSuitesProvider.scala:105 "end": "end"
```